File: tools/gen_buildnum_h.pl
#!/usr/bin/env perl
$|=1;	# Flush writes as soon as print finishes.

use strict;
use warnings;

use File::Basename;

use Cwd;
my $cwd = cwd;

my $scriptpath = dirname($0);

my ($outfile, $prefix, $smprefix) = @ARGV;

die ("Output file not specified") unless(defined $outfile);
die ("Prefix not specified") unless(defined $prefix);
die ("Short prefix not specified") unless(defined $smprefix);

my $tag = "";
my $build = 0;
my $git_tag = "";

if((-e "$scriptpath/../.git") && `git`) {
	$tag = `cd $scriptpath && git describe --tags --abbrev=0 2> /dev/null`;
	if ($tag) {
		chomp($tag);
		$build = `cd $scriptpath && git rev-list --count $tag..HEAD 2> /dev/null`;
		chomp($build);

		$git_tag = `cd $scriptpath && git rev-parse --short HEAD`;
		chomp($git_tag);
		$git_tag = "${tag}-${build}-g${git_tag}";
	}
}

if (!$tag) {
	open RELEASEVER, "$scriptpath/release_ver" || die ("Can't open release_ver:$!");
	$tag = <RELEASEVER>;
	close RELEASEVER;
	chomp($tag);
	$git_tag="$tag";

	if (!$tag) {
		die "error: couldn't get the version information.\n";
	}
}

my @result=split(/[-.]+/, $tag);

my $major = $result[0];
my $minor = $result[1];
my $revis = $result[2];

if( substr($major, 0, 1) eq "v") {
	$major = substr($major, 1);
	$build = $result[3];
}

open FILE, '<', "$scriptpath/build_number.h.in" or die $!;
my @lines = <FILE>;
close FILE or die $!;

foreach(@lines) {
   $_ =~ s/\@SM_PREFIX\@/cc/g;
   $_ =~ s/\@PREFIX\@/CC_LIB/g;
   $_ =~ s/\@VERSION_MAJOR\@/${major}/g;
   $_ =~ s/\@VERSION_MINOR\@/${minor}/g;
   $_ =~ s/\@VERSION_REVISION\@/${revis}/g;
   $_ =~ s/\@VERSION_BUILD\@/${build}/g;
   $_ =~ s/\@GIT_TAG\@/${git_tag}/g;
}

mkdir dirname($outfile);
unlink("$outfile.tmp");

if (-e $outfile) {
	open OUT, '>', "$outfile.tmp" or die $!;
	print OUT @lines;
	close(OUT) or die $!;

	use File::Compare;
	if(compare("$outfile.tmp", "$outfile")) {
		unlink($outfile);
		use File::Copy;
		move "$outfile.tmp", $outfile or die $!;
	} else {
		unlink("$outfile.tmp");
	}
} else {
	print("$outfile\n");
	open(OUT, '>', "$outfile") or die $!;
	print OUT @lines;
	close(OUT) or die $!;
}

File: tools/gen_buildnum_h.py
from filecmp import cmp
from os import path, remove, replace
from shutil import which
from subprocess import run, PIPE
from sys import argv, exit

if len(argv) < 2:
	exit("Error: missing output file")
elif len(argv) < 3:
	exit("Error: missing prefix")
elif len(argv) < 4:
	exit("Error: missing short prefix")

output=argv[1]
prefix=argv[2]
smPrefix=argv[3]

scriptPath = path.dirname(__file__)
tag = ""
build = 0
git_tag = ""

if path.exists(scriptPath + "/../.git") and which("git"):
	result = run('git describe --tags --abbrev=0', cwd=scriptPath, shell=True, stdout=PIPE)
	tag = result.stdout.decode("utf-8").strip("\n\r\t ")

if not tag:
	try:
		with open(scriptPath + '/release_ver', 'r') as file:
			tag = file.read()
		tag = tag.strip("\n\r\t ")
	except EnvironmentError as e: # parent of IOError, OSError *and* WindowsError where available
		print("Can't open release_ver:{0}".format(e))
		raise
else:
	result = run('git rev-list --count {0}..HEAD'.format(tag), cwd=scriptPath, shell=True, stdout=PIPE)
	build = result.stdout.decode("utf-8").strip("\n\r\t ")

	result = run('git rev-parse --short HEAD'.format(tag), cwd=scriptPath, shell=True, stdout=PIPE)
	git_tag = "{0}-{1}-g{2}".format(tag, build, result.stdout.decode("utf-8").strip("\n\r\t "))

result = tag.split(".")

major = None
minor = None
revis = None

if result[0][0] == 'v':
	major = result[0][1:]
	minor = result[1]
	revis = result[2]
	build = result[3]
else:
	major = result[0]
	minor = result[1]
	revis = result[2]

data = None

with open(scriptPath + "/build_number.h.in", "rt") as file:
	data = file.read()

data = data.replace("@SM_PREFIX@", smPrefix)
data = data.replace("@PREFIX@", prefix)
data = data.replace("@VERSION_MAJOR@", major)
data = data.replace("@VERSION_MINOR@", minor)
data = data.replace("@VERSION_REVISION@", revis)
data = data.replace("@VERSION_BUILD@", build)
data = data.replace("@GIT_TAG@", git_tag)

with open(output + ".tmp", "wt") as file:
	file.write(data)

equal = False
try:
	equal = cmp(output + ".tmp", output, shallow=False)
except FileNotFoundError:
	pass

if not equal:
	replace(output + ".tmp", output)
else:
	remove(output + ".tmp")
